### PR TITLE
Adding various array functions

### PR DESCRIPTION
**Pull request by shangyian**
### Summary

Added the following functions:
- `array_distinct`
- `array_except`
- `array_intersect`
- `array_max`
- `array_join`

### Test Plan

Added unit tests for each of the above functions.

- [X] PR has an associated issue: #544
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### CODE REVIEW

**Review by samredai — APPROVED** [review]
Nice! 🚀

File: dj/sql/functions.py [hunk 1]
```diff
@@ -1277,6 +1277,89 @@ def infer_type(  # noqa: F811  # pragma: no cover
...
+
+@ArrayExcept.register
+def infer_type(  # noqa: F811
+    array1: ct.ListType,
+    array2: ct.ListType,
+) -> ct.ListType:
+    return array1.type
+
+
+class ArrayIntersect(Function):  # pylint: disable=abstract-method
```
Comment (samredai) [inline dj/sql/functions.py:1308]: Doesn't have to be done in this PR but I'm realizing we should probably put this `# pylint: disable=abstract-method` once at the top of the file.

Comment (shangyian) [reply dj/sql/functions.py:1308]: Good point! I was going to try to arrange these functions in alphabetical order as well, so I'll do that together in a separate PR

Comment (shangyian) [reply dj/sql/functions.py:1308]: @samredai did that in https://github.com/DataJunction/dj/pull/560

**Pull request merged**